### DISCUSSION TITLE
Cannot use import statement outside a module

### DISCUSSION THREAD

**Discussion by ioquatix** [atom/atom#22891] [Q&A]
Hello, I built and maintain the script runner package: https://github.com/ioquatix/script-runner

I want to update it but I had some problems.

```
Cannot use import statement outside a module
Hide Stack Trace
/Users/samuel/.atom/packages/script-runner/index.js:1
(function (exports, require, module, __filename, __dirname, process, global, Buffer) { return function (exports, require, module, __filename, __dirname) { import './lib/script-runner.mjs'
                                                                                                                                                           ^^^^^^
SyntaxError: Cannot use import statement outside a module
    at new Script (vm.js:84:7)
```

I tried changing the extension of my code to `mjs` and using `import ...` everywhere but I can't seem to get it to load from Atom.

I created an `index.js` file and tried to load the `mjs` file, but I can't get it to work.

Here are the current changes I've made: https://github.com/ioquatix/script-runner/pull/116

Does anyone have any idea how I can fix this issue?

**Comment by icecream17** [comment]
Maybe add `"type": "module"` to package.json?
https://stackoverflow.com/a/60225870/12174015

**Reply by ioquatix** [reply]
I tried that but it didn't seem to help.

**Comment by ioquatix** [comment]
I found this: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

However, I'm unsure whether Atom is compatible with ESM modules, it seems like the version of node is too old?

```
> apm --version
apm  2.6.2
npm  6.14.13
node 12.14.1 x64
atom 1.58.0
python 2.7.18
git 2.32.0
```

**Reply by DeeDeeG** [reply]
Yes, that sounds right.

I vaguely recall that this was due to ECMAScript modules being stabilized surprisingly recently in Node.

This https://nodejs.org/api/esm.html says ECMAScript modules were marked experimental (thus needing a runtime flag on `node` to be enabled, I'm pretty sure) until 12.20.0, and "stabilized" in 12.22.0, whatever that distinction means.

And Atom gets its version of Node from Electron. Efforts to upgrade to newer Electron and therefore newer Node are ongoing, but slow. See: https://github.com/atom/atom/pull/22687

**Reply by DeeDeeG** [reply]
This is not encouraging for official support in Electron: https://github.com/electron/electron/issues/21457
There is a workaround, apparently?: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-import-esm-in-electron. (For example, using the [`esm`](https://github.com/standard-things/esm) package).

**Reply by ioquatix** [reply]
@DeeDeeG thanks for your help in the past. Is there any update on the current situation in Atom 1.60?

**Reply by DeeDeeG** [reply]
Hi again @ioquatix. I am not a huge expert in this, so I checked the electron issue for this again: https://github.com/electron/electron/issues/21457 Seems like people are saying it's still not supported in any version of Electron.

So that leaves workarounds, polyfills, translation layers or build pipelines to convert the code before running it, etc.

I have seen these following workarounds, but I have no idea how good a solution they are. It's a bit over my head honestly.

- https://github.com/standard-things/esm / https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-import-esm-in-electron
- https://github.com/mifi/build-electron

Also, disclaimer: I'm just a contributor to Atom from time to time, so I don't speak as a member of or officially affiliated with the official Atom team. But I'm happy to help out the community and the project here and there when I can!

**Reply by ioquatix** [reply]
Thanks for your help. I tried the `esm` module but ran into other issues unfortunately. I rolled back everything to use  commonjs and pinned all the package versions on commonjs compatible releases, which works.

**Comment by AlexWayfer** [comment]
We have problems with extension for remark, which was updated to ESM: https://github.com/wooorm/linter-remark/issues/14

Probably ESM support by Atom should be an issue rather a discussion.
